### PR TITLE
Add SXG filter extension

### PR DESCRIPTION
**Pull request by cpapazian**
Commit Message: Add SXG filter extension
Additional Description: Filter to transform response to Signed HTTP Exchange (SXG) package using libsxg (https://github.com/google/libsxg). See: https://wicg.github.io/webpackage/draft-yasskin-httpbis-origin-signed-exchanges-impl.html
Risk Level: low, new filter
Testing: unit tests included, filter is based off version running in production for 4 months 
Docs Changes: filter docs added
Release Notes: release notes included

Discussed previously in #9763

### CODE REVIEW

**Comment by cpapazian** [comment]
@rgs1 @twifkak

**Comment by moderation** [comment]
Introduces a new dependency - https://github.com/envoyproxy/envoy/blob/main/DEPENDENCY_POLICY.md#new-external-dependencies. [OSSF Scorecard](https://github.com/ossf/scorecard) results:
```
scorecard --repo=https://github.com/google/libsxg
RESULTS
Active: Pass 10
Automatic-Dependency-Update: Fail 3
Binary-Artifacts: Pass 10
Branch-Protection: Fail 0
CI-Tests: Pass 9
CII-Best-Practices: Fail 10
Code-Review: Pass 10
Contributors: Pass 10
Frozen-Deps: Fail 10
Fuzzing: Fail 10
Packaging: Fail 10
Pull-Requests: Pass 9
SAST: Fail 10
Security-Policy: Fail 5
Signed-Releases: Fail 10
Signed-Tags: Fail 10
Token-Permissions: Fail 10
Vulnerabilities: Pass 10
```

File: CODEOWNERS
Comment (lizan) [inline CODEOWNERS:171]: Anyone from @envoyproxy/maintainers to sponsor?

**Comment by cpapazian** [comment]
@twifkak is it possible to get libsxg to build with cmake 3.10? the arm build is running cmake 3.10.2 currently, and fails: 
```
-- Detecting CXX compile features - done
CMake Error at CMakeLists.txt:67 (cmake_minimum_required):
  CMake 3.13 or higher is required.  You are running version 3.10.2
```
https://dev.azure.com/cncf/envoy/_build/results?buildId=81398&view=logs&j=767be981-567e-57d8-68c3-2140ede0a0bd&t=1fb0b72f-6293-562d-a4d1-bb649223eaca&l=15025

running with the previous commit of libsxg (before we added the 3.13 requirement) we see: 
```
CMake Error at CMakeLists.txt:161 (install):
  install TARGETS given no ARCHIVE DESTINATION for static library target
  "sxg_static".
```
https://dev.azure.com/cncf/envoy/_build/results?buildId=81199&view=logs&j=767be981-567e-57d8-68c3-2140ede0a0bd&t=1fb0b72f-6293-562d-a4d1-bb649223eaca&l=420

not really sure what to make of this. would be curious your thoughts ...

**Comment by twifkak** [comment]
@cpapazian Yep, I'll send a PR to update the minimum requirement to 3.1 (earliest I could test easily on my x86_64 laptop via https://cmake.org/files/) when RUN_TEST is off.

It looks like the subsequent error can be fixed by adding the flags `-DCMAKE_INSTALL_LIBDIR=lib -DCMAKE_INSTALL_BINDIR=bin`. I guess those became the default in some cmake 3.11+ version? Anyway, I'll just include those flags in the README of the same PR.

**Comment by twifkak** [comment]
@cpapazian Merged google/libsxg#95 which hopefully fixes this (with tests to prevent regression).

File: api/envoy/extensions/filters/http/sxg/v3alpha/sxg.proto
Comment (rgs1) [inline api/envoy/extensions/filters/http/sxg/v3alpha/sxg.proto:27]: can you comment a bit more on how these two will be used? Maybe with a link to the SXG docs explaining how it's used, etc.

File: api/envoy/extensions/filters/http/sxg/v3alpha/sxg.proto
Comment (rgs1) [inline api/envoy/extensions/filters/http/sxg/v3alpha/sxg.proto:29]: Is that default an SXG recommendation? If so, add link please.

File: api/envoy/extensions/filters/http/sxg/v3alpha/sxg.proto
Comment (rgs1) [inline api/envoy/extensions/filters/http/sxg/v3alpha/sxg.proto:34]: More info and/or link to SXG docs.

File: api/envoy/extensions/filters/http/sxg/v3alpha/sxg.proto
Comment (rgs1) [inline api/envoy/extensions/filters/http/sxg/v3alpha/sxg.proto:38]: We should either add validation here or do in code and a unit test for that.

File: api/envoy/extensions/filters/http/sxg/v3alpha/sxg.proto
Comment (rgs1) [inline api/envoy/extensions/filters/http/sxg/v3alpha/sxg.proto:43]: Is this required? If so we need `[(validate.rules).string = {min_bytes: 1}]`. Otherwise comment and say what the default is.

File: api/envoy/extensions/filters/http/sxg/v3alpha/sxg.proto
Comment (rgs1) [inline api/envoy/extensions/filters/http/sxg/v3alpha/sxg.proto:55]: ditto for all of these

File: docs/root/configuration/http/http_filters/sxg_filter.rst
Comment (rgs1) [inline docs/root/configuration/http/http_filters/sxg_filter.rst:24]: let's RST link that link

File: docs/root/configuration/http/http_filters/sxg_filter.rst
Comment (rgs1) [inline docs/root/configuration/http/http_filters/sxg_filter.rst:53]: I'd remove the comments from the config example

File: source/extensions/extensions_metadata.yaml
Comment (rgs1) [inline source/extensions/extensions_metadata.yaml:274]: I think we actually want `robust_to_untrusted_downstream`, given we need to trust the upstream to sign its payload no?

File: source/extensions/filters/http/sxg/filter.h
Comment (rgs1) [inline source/extensions/filters/http/sxg/filter.h:92]: should these return values be references since that's what the secret_reader_ methods return?

File: docs/root/configuration/http/http_filters/sxg_filter.rst
Comment (rgs1) [inline docs/root/configuration/http/http_filters/sxg_filter.rst:65]: `*<stat_prefix>.sxg.*` namespace

File: source/extensions/filters/http/sxg/filter.cc
Comment (rgs1) [inline source/extensions/filters/http/sxg/filter.cc:50]: nit you can drop the () around the comparison

File: source/extensions/filters/http/sxg/filter.cc
Comment (rgs1) [inline source/extensions/filters/http/sxg/filter.cc:53]: ditto

File: source/extensions/filters/http/sxg/filter.cc
Comment (rgs1) [inline source/extensions/filters/http/sxg/filter.cc:92]: nit drop

**Review by rgs1 — CHANGES_REQUESTED** [review]
I still need to review filter_test, but left a few comments to get it going. Thanks!

File: api/envoy/extensions/filters/http/sxg/v3alpha/sxg.proto
Comment (markdroth) [inline api/envoy/extensions/filters/http/sxg/v3alpha/sxg.proto:25]: Does the cert info need to come from SDS?  We've recently been adding other options to `CommonTlsContext`, such as the `CertificateProviderInstance` hooks.  Can we leverage that here somehow instead of hard-coding use of SDS?

Comment (cpapazian) [reply api/envoy/extensions/filters/http/sxg/v3alpha/sxg.proto:25]: No, it does not need to be SDS. The version of this that we are running currently, uses an internal platform for secrets. I was using the Oauth2 filter extension as a template, and it uses SDS. Happy to use `CertificateProviderInstance` or similar if that's preferred. Do you have an example I could look at?

File: api/envoy/extensions/filters/http/sxg/v3alpha/sxg.proto
Comment (cpapazian) [reply api/envoy/extensions/filters/http/sxg/v3alpha/sxg.proto]: @twifkak what do you think? is there any documentation of recommended values for duration?

File: api/envoy/extensions/filters/http/sxg/v3alpha/sxg.proto
Comment (cpapazian) [reply api/envoy/extensions/filters/http/sxg/v3alpha/sxg.proto]: thoughts on the validation? would it be enough to validate to ensure a value is present? or should i validate for the actual contents of the value?

File: api/envoy/extensions/filters/http/sxg/v3alpha/sxg.proto
Comment (rgs1) [reply api/envoy/extensions/filters/http/sxg/v3alpha/sxg.proto:26]: What's the rationale for departing from SDS?

Comment (rgs1) [reply api/envoy/extensions/filters/http/sxg/v3alpha/sxg.proto:26]: @envoyproxy/api-shepherds to get some guidance on SDS vs less SDS for similar use cases.

Comment (markdroth) [reply api/envoy/extensions/filters/http/sxg/v3alpha/sxg.proto:26]: Unfortunately, the way the protos are currently set up, we don't seem to have a single reusable proto that represents a "certificate source" that can be configured for any of the various supported mechanisms.  The infrastructure here is a little weak.

For now, I guess this is okay as-is, just because we don't have anything better to offer.  Just keep in mind that we will probably need to update this in the future as this infrastructure improves.

**Review by markdroth — COMMENTED** [review]
/lgtm api

File: api/envoy/extensions/filters/http/sxg/v3alpha/sxg.proto
Comment (twifkak) [reply api/envoy/extensions/filters/http/sxg/v3alpha/sxg.proto]: There isn't any canonical documentation for this (it's on my to-do list). My short answer is 1 day if the resource is HTML served without a CSP header or if it is JS (regardless of CSP header), and 7 days otherwise. If differentiating by content type is infeasible, then I would stick to a default of 1 day. But it's not a super strong recommendation.

In your case, I would either a) get rid of the 1-day backdating or b) make the duration 2 days so that it's effectively 1 day from now. I'm torn between these two options; either is fine.

### Details re: duration

What's on my mind is a trade-off between defense-in-depth and speed:

1. If a security vulnerability is discovered in the client-side code, an attacker could ask the server for an SXG and then redistribute it on its own serving path (e.g. via a social media campaign), thereby prolonging the attack for up to `$duration` days after the frontend is fixed.
   - Service workers, having a similar concern, [limit cache duration to 24 hours](https://chromium.googlesource.com/chromium/src/+/main/docs/security/service-worker-security-faq.md#if-a-site-has-an-xss-vulnerability_can-the-attacker-permanently-compromise-that-origin-for-me).
   - JS is the predominant client-side vector for vulnerabilities.
   - CSP headers can mitigate the risks.

2. For SXGs used for prefetching from referrers (the typical use case), a longer duration means a better cache hit rate, which increases its LCP benefit.
   - IME, 1 day is a reasonable "shelf" for the curve of cache hit rate vs duration. Lower than that will cause a more significant drop.

Thus, my default recommendation is 7 day duration for everything except a 1 day duration for:
- JavaScript
- HTML with inline JavaScript without a CSP header

There's logic for that [in AMP Packager](https://github.com/ampproject/amppackager/blob/8ae2d36d8786c5f15c43ca91c5d171532a8185c4/transformer/transformer.go#L288-L315) and in [Web Packager](https://github.com/google/webpackager/blob/6af4d5194ad2f6bc5daf68293f07cc25ba100126/server/from_toml.go#L150-L162) ([here too](https://github.com/google/webpackager/blob/c7597c58f5e83513153ff8b77ad2e7486bd2c597/processor/htmlproc/htmltask/sub_content_type.go#L51-L59)). (The CSP header condition is not implemented; that was a recent idea a colleague came up with.)

Since parsing HTML is probably too heavyweight a dependency for an Envoy extension, I'd assume all HTML contains inline JS, hence my short answer.

### Details re: backdating

 - Option (a) slightly reduces the LCP benefit of prefetching because clients with badly skewed clocks will reject the SXG as too new, but IME the impact of this is <1%, so maybe this an acceptable loss?
 - Option (b) fixes this, but slightly worsens the security posture for those <1% of clients.

Comment (cpapazian) [reply api/envoy/extensions/filters/http/sxg/v3alpha/sxg.proto]: Interesting. So, perhaps it makes sense to allow for downstream services to specify a duration that the envoy filter can consume. I'll update the docs to suggest 2 days for javascript (1 day, when accounting for the backdating) and 7 days for everything else, with a note about inline JS. Thanks!

Comment (cpapazian) [reply api/envoy/extensions/filters/http/sxg/v3alpha/sxg.proto]: I can also make hte backdating timespan configurable. let me know what you think makes sense here, and i can update. in the meantime i've updated the notes: ```
  // The duration for which the generated SXG package will be valid. Default is 604800s (7 days in seconds).
  // Note that in order to account for clock skew, the timestamp will be backdated by a day. So, if duration
  // is set to 7 days, that will be 7 days from 24 hours ago (6 days from now). Also note that while 6/7 days 
  // is appropriate for most content, if the downstream service is serving Javascript, or HTML with inline 
  // Javascript, 1 day (so, with backdated expiry, 2 days, or 172800 seconds) is more appropriate.
```

Comment (twifkak) [reply api/envoy/extensions/filters/http/sxg/v3alpha/sxg.proto]: The documentation change sounds good (and sufficient) to me.

I'm happy to defer to your opinion on making backdating configurable. I'm hopeful that there's a reasonable default that allows most people not to have to think about it. But if it's cheap to implement and maintain, then it seems nice to have.

For the future version where downstream services specify duration, then maybe something like `min(7 days, [s-maxage or max-age from cache-control header] + $backdate_amount)` would work?

File: source/extensions/filters/http/sxg/filter.cc
Comment (rgs1) [inline source/extensions/filters/http/sxg/filter.cc:376]: should we move some of these utility methods into their own file (e.g.: `sxg_utils.cc`) to keep this file cleaner/shorter?

File: source/extensions/filters/http/sxg/filter.cc
Comment (twifkak) [reply source/extensions/filters/http/sxg/filter.cc:170]: I'm not familiar with Envoy but it looks like `Buffer::Instance::move()` takes another `Buffer::Instance`. (I don't see an easy way to construct that from a `string_view`, but I may have missed it.)

Perhaps [`addBufferFragment()`](https://github.com/envoyproxy/envoy/blob/6834081f835d609606e77ef84d734e30fc403d58/envoy/buffer/buffer.h#L150-L155) could be used? It looks like `enc_buf` needs to outlive this lambda, so I think you'll need to allocate the `sxg_buffer_t` on the heap and then construct a [`BufferFragmentImpl`](https://github.com/envoyproxy/envoy/blob/0fae6970ddaf93f024908ba304bbd2b34e997a51/source/common/buffer/buffer_impl.h#L598) whose `releasor` calls something like `sxg_buffer_release(result); delete result;`. Does that sound right?

I think whether you can can replace `sxg_buffer_release` with a call to `free(result.data)` depends on whether `OPENSSL_free` has been overridden ([BoringSSL](https://github.com/google/boringssl/blob/c1571feb5faf5cce844354c63d0f3e842464bea3/crypto/mem.c#L155-L166); [OpenSSL](https://github.com/openssl/openssl/blob/1c0eede9827b0962f1d752fa4ab5d436fa039da4/crypto/mem.c#L247-L250)).

Comment (cpapazian) [reply source/extensions/filters/http/sxg/filter.cc:170]: `addBufferFragment(...)` is working for me locally. `result.data` is already allocated on the heap, right? so if it's safe to call `OPENSSL_free` on `result.data` directly (rather than calling `sxg_buffer_release`), then we're good. if not, we can either allocate the `sxg_buffer_t` on the heap or add it to the filter class.

Comment (twifkak) [reply source/extensions/filters/http/sxg/filter.cc:170]: Ah yeah, you're right; we don't need to move the `sxg_buffer_t` to the heap. I guess either option works:
 1. Capture the `sxg_buffer_t` by value into the `releasor` lambda, and use the standard `sxg_buffer_release`.
 2. Capture the `uint8_t*` by value and use `OPENSSL_free`.

(1) seems slightly nicer since it adheres to the higher-level [documented contract](https://github.com/google/libsxg/blob/672b807e9a1a0032c73f1eeb7c9761d8cfd013d0/include/libsxg/sxg_buffer.h#L39-L40) in sxg_buffer.h, so less likely to break with a dependency update. But if that's suboptimal for some other reason (performance?) then (2) works for me, too. In that case, maybe send a PR to update the comment in sxg_buffer.h so it becomes a part of the contract.

**Comment by cpapazian** [comment]
@yanavlasov I think this is ready for review. Let me know if there's anything else you need from me first. Thanks!

cc @rgs1

**Comment by cpapazian** [comment]
per @rgs1, we now have a contrib directory (https://github.com/envoyproxy/envoy/pull/17595). I can move this extension there, if that makes sense. let me know which you prefer, @yanavlasov

**Comment by junr03** [comment]
@yanavlasov friendly ping on this.

**Comment by cpapazian** [comment]
@yanavlasov - per offline conversation with @rgs1, i went ahead and moved this over to contrib.

**Comment by rgs1** [comment]
@mattklein123 does this sound like a reasonable candidate for contrib/? We were initially targeting it as a regular filter, but contrib/ would be great for us so we can share the code with other parties and then possibly graduate out of contrib... Thanks!

**Comment by mattklein123** [comment]
> @mattklein123 does this sound like a reasonable candidate for contrib/? We were initially targeting it as a regular filter, but contrib/ would be great for us so we can share the code with other parties and then possibly graduate out of contrib... Thanks!

Sure! Is this ready to review? I can take a look and merge if so.

**Comment by cpapazian** [comment]
> Sure! Is this ready to review? I can take a look and merge if so.

Yeah, I think we're ready. @rgs1 has already taken a pass at review.

Only question we had is what to do with the [libsxg](https://github.com/google/libsxg/) dependency now that the filter has moved to contrib. It wasn't obvious to me how to dis-entangle all of the external repo stuff in the builds, and I'm not sure we even want to. 

Thanks!

File: docs/root/version_history/current.rst
Comment (mattklein123) [inline docs/root/version_history/current.rst:99]: note this is available in contrib builds only with a ref link to the contrib docs.

File: docs/root/configuration/http/http_filters/sxg_filter.rst
Comment (mattklein123) [inline docs/root/configuration/http/http_filters/sxg_filter.rst:8]: Can you note this is in contrib builds only with a ref link to the contrib docs. If the squash filter docs don't have that can you fix that also? (I know I automated the other docs but I think for these manual docs it is probably missing.)

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks I'm fine with this going in for contrib for now, with a few small doc comments.

Regarding the build dependencies being in core, in a perfect world we would figure out how to have them live with the contrib extension itself but I don't think we have a good pattern for that so I'm OK with this for now, especially since bazel will only build the dependency if it's actually used.

/wait

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**
